### ISSUE TITLE
Performance issue

### ISSUE THREAD

**Issue by uxcn** [sahib/rmlint#168]
I've been running some benchmarks for different deduplicators, and I ran into either a performance issue or a bug using `rmlint` on a particular setup.  The operating system is Linux 4.3.2 (BFQ) and the disk is an LVM logical volume over an md RAID5 array (EXT4).  The version for `rmlint` is 2.4.2.

Running `rmlint`, the software effectively times out (1hr+) with the error message...

`WARNING: read buffer limit reached - waiting for processing to catch up`

I was able to successfully test `rmlint` on the same host using a different individual disk though (LVM+EXT4).

**Comment by SeeSpotRun** [comment]
The warning just means that files are being read faster than the CPU can calculate the SHA1 checksums.

Are you sure it's hung?  Check ```$ top``` output; if rmlint still has non-zero CPU usage then it's still running.  Also try running with the progressbar option (eg ```-g``` or ```-o progressbar```) and see if rmlint is still reporting progress.

**Comment by uxcn** [comment]
I'm not sure if it's completely locked up.  The RAID array has an uncached read throughput of 200+ MiB/s, and around 2 TiB of data though.  I let it run for 1hr+, so it shouldn't be an issue with computing SHA-1.  It might be a threading problem or an i/o problem or it might be a kernel bug.

I can run it with debug symbols built in and get a core file if it helps.

**Comment by uxcn** [comment]
I ran `rmlint` with `--progress`.  Here's a snapshot of the results...

```
```

So, it's reading over 1 TiB from disk.

**Comment by uxcn** [comment]
I ran `rmlint` with `--progress`.  It isn't completely locking up, but I did notice something else odd...

```
jason@ru ~ rmlint --hidden -o fdupes:/dev/null --progress /multimedia            
▕░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▏                                                                                                                                                               Traversing (10269 usable files / 0 + 0 ignored files / folders)
▕░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▏                                                                                                                                                                  Preprocessing (reduces files to 2409 / found 112 other lint)
▕▒░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░C4;01m░░░░░░░░░░░░░░░░░░░░░░░░░░░▏                                                                                                                                                         Matching (24 dupes of 11 originals; 1053.96 GB to scan in 1079 fi
```

Ir's reading more than 1 TiB from disk.  Am I reading the output right?

**Comment by SeeSpotRun** [comment]
Correct; there must be some large files that are at least partial duplicates.  

If cpu is the bottleneck then maybe run with a lighter checksum eg ```-a city256```.

**Comment by SeeSpotRun** [comment]
Out of interest, do other dupe finders also struggle with the same data set?

Could perhaps be related to #166 if you have some large hard linked dupes.

**Comment by uxcn** [comment]
Two deduplicators had issues with this particular benchmark (`duff` was the other one).

I read through #166 and tested again using `--no-hardlinked` which finishes an order of magnitude faster (2m05s).  Is there a reason you're considering hardlinks as potential duplicates by default?

**Comment by SeeSpotRun** [comment]
No, we just haven't implemented the logic to identify when to read and when not (ie if all files in the group are hardlinks to the same file).  It's a little trickier than it sounds if we want to cover corner cases (eg a group of 3 file of which the first 2 are hardlinks and the 3rd one is identical for the first MB then diverges).
Anyway try now with https://github.com/SeeSpotRun/rmlint/tree/feature/skip_hardlinks

**Comment by uxcn** [comment]
I think I'm a little confused.  Isn't `--no-hardlinked` treating identical inode's as non-duplicate?  Are files not being grouped by device and inode first?

The `--no-hardlinked` performance looks roughly like what I would expect.

**Comment by SeeSpotRun** [comment]
Default behaviour is to consider hardlinks as separate files and identify them as duplicates in the output.  For example:


> $ # make some hardlinked duplicates:
$ mkdir test && cd test
$ echo data > file1 && echo data > file2
$ ln file1 link1 && ln file2 link2
$ # scan them:
$ rmlint
> 
\# Duplicate(s):
    **ls** '/home/daniel/GIT/rmlint_ssr/test/file2'
    **rm** '/home/daniel/GIT/rmlint_ssr/test/file1'
    **rm** '/home/daniel/GIT/rmlint_ssr/test/link2'
    **rm** '/home/daniel/GIT/rmlint_ssr/test/link1'
==> Note: Please use the saved script below for removal, not the above output.
==> In total 4 files, whereof 3 are duplicates in 1 groups.
==> This equals 5 B of duplicates which could be removed.`

With ```--no-hardlinked``` the existing hardlinks are ignored:
> $rmlint --no-hardlinked
>
\# Duplicate(s):
    **ls** '/home/daniel/GIT/rmlint_ssr/test/file2'
    **rm** '/home/daniel/GIT/rmlint_ssr/test/file1'
==> Note: Please use the saved script below for removal, not the above output.
==> In total 4 files, whereof 1 are duplicates in 1 groups.
==> This equals 5 B of duplicates which could be removed.

Depending on output settings, the rmlint-generated `rmlint.sh` script can either delete (default) or hardlink (`-c sh:hardlink`) the identified duplicates:

> $ # display inodes:
$ stat -c "%i %n" * | sort
8133683 file1
8133683 link1
8133684 file2
8133684 link2
$ # link duplicates using rmlint
$ rmlint -c sh:hardlink
...
$ ./rmlint.sh
...
Keeping:  /home/foo/test/file2
Hardlinking to original: /home/foo/test/file1
Leaving as-is (already hardlinked to original): /home/foo/test/link2
Hardlinking to original: /home/foo/test/link1
$ # display inodes:
$ stat -c "%i %n" * | sort
8133684 file1
8133684 file2
8133684 link1
8133684 link2
>
> $ # run rmlint again in default mode to delete duplicates:
$ rmlint
...
$ ./rmlint.sh
Keeping:  /home/foo/test/file2
Deleting: /home/foo/test/link2
Deleting: /home/foo/test/file1
Deleting: /home/foo/test/link1

**Comment by uxcn** [comment]
The `--no-hardlinked` behavior is what I think most people would expect (including me).

With POSIX filesystems, the device id and inode should uniquely identify the actual data.  So, if you group by them first, you shouldn't run the risk of reading blocks more than once.  Otherwise, there are pathological scenarios like the above and #166.

If you're curious, there's a simple example [here](https://github.com/uxcn/yafd/blob/master/src/c/worker.c).

**Comment by SeeSpotRun** [comment]
It's not always easy to define what should be the intuitive default behaviour.  If the user's intention is to declutter their filesystem tree then they will generally want to find and delete the hardlinks.  If then intention is to save space then they won't care.

The following appear to ignore hardlinks by default: yafd, fdupes, rdfind.
On the other hand, fslint/findup reports all hardlinks, provided they have at least one duplicate with a different inode.

So it looks like rmlint is a bit out of step - perhaps we should change the default behaviour.


Regarding pathological behavior, actually rmlint never reads the same block more than once because it already groups files according to inode number.  What it was doing however that was not smart is illustrated as follows:
> $ # make a big file and a hardlink thereof
$ truncate -s 1G file1
$ ln file1 link1
$ /usr/bin/time rmlint file1 link1
...
4.01user 0.74system 0:04.09elapsed 116%CPU (0avgtext+0avgdata 231240maxresident)k
$
$ # how prove that it only read the data once?
$ truncate -s 1G file2
$ /usr/bin/time rmlint file1 file2
...
8.21user 2.31system 0:04.21elapsed 250%CPU (0avgtext+0avgdata 374364maxresident)k

What's dumb in the first example above is that rmlint didn't need to read the data at all, since there were only 2 files and they were already hardlinked.  This is fixed as of https://github.com/SeeSpotRun/rmlint/commit/3bdd2504609662ae8e8f62a8d3c6a2ab6b7832ec.  The new timing for the first example is:
> $ rmlint_3bdd250 file1 link1
...
0.00user 0.00system 0:00.07elapsed 18%CPU (0avgtext+0avgdata 5560maxresident)k

**Comment by uxcn** [comment]
> It's not always easy to define what should be the intuitive default behaviour. If the user's intention is to declutter their filesystem tree then they will generally want to find and delete the hardlinks. If then intention is to save space then they won't care.

It's definitely more flexible to be able to do both (`yafd` currently doesn't).  I tend to use mine for multimedia, which is what I'd guess is one of the bigger things people use deduplicators for (e.g. pictures, movies, music, etc...).  Nobody else is using `yafd` yet, AFAIK.  I was planning on doing an alpha release for people to test, which was when I noticed the `rmlint` performance issue.  Obviously, I wanted to let you guys know since it could have affected your users.

> What's dumb in the first example above is that rmlint didn't need to read the data at all, since there were only 2 files and they were already hardlinked. This is fixed as of SeeSpotRun@3bdd250. The new timing for the first example is:
> $ rmlint_3bdd250 file1 link1
> ...
> 0.00user 0.00system 0:00.07elapsed 18%CPU (0avgtext+0avgdata 5560maxresident)k

Those numbers are definitely more like what I would have expected to see.  I haven't looked at the `rmlint` algorithm, but I'm guessing the default is `O(nk)`.

Also just a heads up, I noticed there isn't a Gentoo/Funtoo ebuild for `rmlint` (I have to build manually).

**Comment by sahib** [comment]
> Also just a heads up, I noticed there isn't a Gentoo/Funtoo ebuild for rmlint (I have to build manually).

Actually, [there](https://github.com/Jormangeud/jrg-overlay/tree/master/app-misc/rmlint) are [ebuilds](https://github.com/Bfgeshka/octopus/blob/master/app-misc/rmlint/rmlint-9999.ebuild) from users in Issue #143. As Gentoo user you'll probably want to support [this ebuild request](https://bugs.gentoo.org/show_bug.cgi?id=560390) (I guess that's what necessary to get in the official tree).

**Comment by uxcn** [comment]
I voted on the bug.  I've actually had more luck submitting pull requests to the Gentoo [github](https://github.com/gentoo/gentoo) repository (Funtoo also depends on portage).  There are a _lot_ of packages missing (`cling`, `libc++`, `urxvt 24bit`. etc...).  Honestly, I might not have written a deduplicator if `rmlint` was already in the repository.

**Comment by sahib** [comment]
Thanks for the hint. How would a pull request look like there in detail? We're not familiar with neither Gentoo, nor it's packaging procedure. I found [this](https://github.com/gentoo/gentoo/pull/437) old PR of your's and thought you could maybe give us a bit of starter help?

**Comment by uxcn** [comment]
I took a look at the ebuilds, and they generally look okay.  The one thing I did notice was missing `USE` flags (`IUSE` in the ebuilds).  I'm not extremely familiar with `rmlint`, but any features that might make sense for people to be able to configure at compile time are a good idea to make configurable via.  For example, if a feature depends on a library, people concerned with footprint might want to compile it out (e.g. embedded users).

The other thing I noticed was that `CFLAGS` don't seem to be passed down to SCons.  There are some packages that strip `CFLAGS` by default now (e.g. Chrome, Firefox) but even following that route it's a good idea to add an escape hatch for people that are doing things like cross-compiling.  The `custom-cflags` variable is the de-facto flag to enable them.  Most packages enable `CFLAGS` by default though.

As to how the actual pull requests should look, it's generally one commit per *change*.  If there's a bug report, it's a good idea to reference it via `Gentoo-Bug`.  It's probably also a good idea to add a `Suggested-By`.  As to the comment, I generally try to make as much sense as possible assuming zero context (e.g. "fixes segfault" would be bad).    If anybody's curious about the ebuild syntax, the [dev manual](https://devmanual.gentoo.org/) is a fairly decent reference.

The important thing to keep in mind about Gentoo is that it strives to be as flexible as possible for users, so configurability is extremely important.  If you guys have anymore questions, I can try to answer as time permits.  Unfortunately, I'm a bit busy as of late.

I hope this helps.

**Comment by sahib** [comment]
I'll close this, since the original issue seems to be addressed.
Discussion/Updates regarding packaging should happen in #143.
